A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+other values. We considered the alternative of NACKing (i.e. rejecting)
+the CDS update when unsupported values are present, however it was discarded
+because of the difficulty in using such configuration in mixed deployments.
+
+The `override_host_status` value is included in the config update sent to
+the `“cluster_resolver_experimental”` policy which passes down the update to
+the `“xds_cluster_impl”` policy where it is used to create the additional
+child policy as described below.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing and just pass the RPC through (to the next
+filter). For example if the `path` configuration value is `/Package1.Service2`
+and the RPC method is `/Package2.Service1/Method3` then just pass the RPC
+through.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
+warning and pass the RPC through. For a valid `override-host` value set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path of the RPC, if the RPC path does not match the
```

Comment:
The `RPC path match` has already occurred in the first step (as part of the request) so doesn't need to be repeated here, correct?

Reply:
It does. The filter has been set on the channel so all requests and responses will pass through the filter. So in the response (return) path I still need to skip processing of a response if the "path" of the RPC didn't match the cookie config. Hope you agree?

Reply:
I think this is an implementation detail that doesn't really need to be specified here.  In C-core, we will probably not need to duplicate the path-match check, because we will simply not intercept the recv_trailing_metadata callback if the path doesn't match on the sending side.

Reply:
@markdroth that's what I was trying to hint at.

Reply:
@markdroth at least in Java the filter is set on the channel and is invoked in both the directions. The code inside the filter decides to skip processing if path doesn't match. On the request (outbound) side it can set a boolean which can be checked on the return path to skip processing.